rio/src/RIO/File.hs

```diff
@@ -0,0 +1,349 @@
+foreign import ccall unsafe "HsBase.h __hscore_o_rdwr"   o_RDWR   :: CInt
+foreign import ccall unsafe "HsBase.h __hscore_o_append" o_APPEND :: CInt
+foreign import ccall unsafe "HsBase.h __hscore_o_creat"  o_CREAT  :: CInt
+foreign import ccall unsafe "HsBase.h __hscore_o_noctty" o_NOCTTY :: CInt
+
+-- After here, we have our own imports
+foreign import ccall safe "fcntl.h openat"
+  c_safe_openat :: CInt -> CFilePath -> CInt -> CMode -> IO CInt
+
+foreign import ccall safe "fcntl.h renameat"
+  c_safe_renameat :: CInt -> CFilePath -> CInt -> CFilePath -> IO CInt
+
+foreign import ccall safe "unistd.h fsync"
+  c_safe_fsync :: CInt -> IO CInt
+
+std_flags, output_flags, read_flags, write_flags, rw_flags,
+    append_flags :: CInt
+std_flags    = o_NOCTTY
+output_flags = std_flags    .|. o_CREAT
+read_flags   = std_flags    .|. o_RDONLY
+write_flags  = output_flags .|. o_WRONLY
+rw_flags     = output_flags .|. o_RDWR
+append_flags = write_flags  .|. o_APPEND
+
+ioModeToFlags :: IOMode -> CInt
+ioModeToFlags iomode =
+  case iomode of
+    ReadMode      -> read_flags
+    WriteMode     -> write_flags
+    ReadWriteMode -> rw_flags
+    AppendMode    -> append_flags
+
+openDir :: MonadIO m => FilePath -> m Fd
+openDir fp
+  -- TODO: Investigate what is the situation with Windows FS in regards to non_blocking
+  -- NOTE: File operations _do not support_ non_blocking on various kernels, more
+  -- info can be found here: https://ghc.haskell.org/trac/ghc/ticket/15153
+ =
+  liftIO $
+  withFilePath fp $ \cFp ->
+    Fd <$>
+    (throwErrnoIfMinus1Retry "openDir" $
+     c_safe_open cFp (ioModeToFlags ReadMode) 0o660)
+
+closeDirectory :: MonadIO m => Fd -> m ()
+closeDirectory (Fd dirFd) =
+  liftIO $
+  void $
+  throwErrnoIfMinus1Retry "closeDirectory" $ c_close dirFd
+
+fsyncFileDescriptor :: MonadIO m => String -> CInt -> m ()
+fsyncFileDescriptor name cFd =
+  liftIO $
+  void $
+    throwErrnoIfMinus1 ("fsync - " <> name) $
+    c_safe_fsync cFd
+
+-- NOTE: Not Async exception safe
+openFileFromDir :: (MonadIO m) => Fd -> FilePath -> IOMode -> m Handle
+openFileFromDir (Fd dirFd) fp iomode =
+  liftIO $
+  withFilePath fp $ \f -> do
+    bracketOnError
+      (throwErrnoIfMinus1Retry "openFileFromDir" $
+       c_safe_openat dirFd f (ioModeToFlags iomode) 0o666)
+      (void . c_close)
+      (\fileFd -> do
+         (fD, fd_type) <-
+           FD.mkFD
+             fileFd
+             iomode
+             Nothing {- no stat -}
+             False {- not a socket -}
+             False {- non_blocking --}
+         -- we want to truncate() if this is an open in WriteMode, but only if the
+         -- target is a RegularFile. ftruncate() fails on special files like
+         -- /dev/null.
+         when (iomode == WriteMode && fd_type == RegularFile) $
+           Device.setSize fD 0
+         HandleFD.mkHandleFromFD fD fd_type fp iomode False Nothing
+           `onException` (liftIO $ Device.close fD))
```

Comment:
I'm not sure about this bit. We already have a `c_close` error handling installed via the `bracketOnError`. Won't we end up with a double close here via the `Device.close` call? And if not, why don't we need a `onException` bit for the `setSize` call?

Reply:
Argh, great catch, I'll refactor this so that the `mkFD` is part of the open subroutine of the `bracket` call

Reply:
Nice setup. On a separate note, we may want to save this code as an example of where Rust is far easier to write correctly than Haskell.